packages/core-app-api/src/apis/implementations/FetchApi/BackstageProtocolResolverFetchMiddleware.ts

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright 2021 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { FetchFunction, FetchMiddleware } from './types';
+
+function join(left: string, right: string): string {
+  if (!right || right === '/') {
+    return left;
+  }
+
+  return `${left.replace(/\/$/, '')}/${right.replace(/^\//, '')}`;
+}
+
+/**
+ * Handles translation from backstage://some-plugin-id/<path> to concrete
+ * http(s) URLs.
+ *
+ * @public
+ */
+export class BackstageProtocolResolverFetchMiddleware
+  implements FetchMiddleware
+{
+  constructor(
+    private readonly discovery: (pluginId: string) => Promise<string>,
+  ) {}
+
+  /**
+   * {@inheritdoc FetchMiddleware.apply}
+   */
+  apply(next: FetchFunction): FetchFunction {
+    return async (input, init) => {
+      const request = new Request(input, init);
+      const { protocol, hostname, pathname, search, hash, username, password } =
+        new URL(request.url);
+
+      if (protocol !== 'backstage:') {
+        return next(input, init);
+      }
+
+      let base = await this.discovery(hostname);
+      if (username || password) {
```

Comment:
This behaves differently in the browsers. Tbh I think it should be safe to skip support for user/pass auth, at least as passed to fetch, but it could probably be returned from the discovery API

Reply:
How do you mean it behaves differently? 🤔  Testing with `new URL` in the browser console does seem to get the expected fields returned. I thought it might be unexpected for a caller to have their username/password stripped out without warning.

Reply:
Try for example these two in the browser:

```js
new URL('http://a:b@example.com').username
new URL('backstage://a:b@example.com').username
```

Reply:
Ahh right I forgot! Good eye. Damn, a lot of other stuff is completely broken too, `pathname` becomes `//a:b@example.com`. Let me give that a look...

Reply:
Alright, fixed - but with retained username/password handling :)